crates/uv-git/src/git.rs

```diff
@@ -206,266 +242,151 @@ impl GitRemote {
-        let checkout = match git2::Repository::open(destination)
+        let checkout = match GitRepository::open(destination)
             .ok()
-            .map(|repo| GitCheckout::new(self, rev, repo))
+            .map(|repo| GitCheckout::new(rev, repo))
             .filter(GitCheckout::is_fresh)
         {
             Some(co) => co,
             None => GitCheckout::clone_into(destination, self, rev)?,
         };
-        checkout.update_submodules(strategy, client)?;
+        checkout.update_submodules()?;
         Ok(checkout)
     }
 
     /// Get a short OID for a `revision`, usually 7 chars or more if ambiguous.
-    pub(crate) fn to_short_id(&self, revision: git2::Oid) -> Result<GitShortID> {
-        let obj = self.repo.find_object(revision, None)?;
-        Ok(GitShortID(obj.short_id()?))
+    pub(crate) fn to_short_id(&self, revision: GitOid) -> Result<String> {
+        let output = ProcessBuilder::new("git")
+            .arg("rev-parse")
+            .arg("--short")
+            .arg(revision.as_str())
+            .cwd(&self.repo.path)
+            .exec_with_streaming(&mut silent, &mut silent, true)?;
+
+        let mut result = String::from_utf8(output.stdout)?;
+        result.truncate(result.trim_end().len());
+        Ok(result)
     }
 
     /// Checks if the database contains the object of this `oid`.
-    pub(crate) fn contains(&self, oid: git2::Oid) -> bool {
-        self.repo.revparse_single(&oid.to_string()).is_ok()
+    pub(crate) fn contains(&self, oid: GitOid) -> bool {
+        self.repo.rev_parse(oid.as_str()).is_ok()
     }
 }
 
 impl GitReference {
     /// Resolves self to an object ID with objects the `repo` currently has.
-    pub(crate) fn resolve(&self, repo: &git2::Repository) -> Result<git2::Oid> {
+    pub(crate) fn resolve(&self, repo: &GitRepository) -> Result<GitOid> {
         let refkind = self.kind_str();
-        let id = match self {
-            Self::Tag(s) => (|| -> Result<git2::Oid> {
-                let refname = format!("refs/remotes/origin/tags/{s}");
-                let id = repo.refname_to_id(&refname)?;
-                let obj = repo.find_object(id, None)?;
-                let obj = obj.peel(ObjectType::Commit)?;
-                Ok(obj.id())
-            })()
-            .with_context(|| format!("failed to find tag `{s}`"))?,
-
-            Self::Branch(s) => {
-                let name = format!("origin/{s}");
-
-                // Resolve the remote name since that's all we're configuring in
-                // `fetch` below.
-                repo.find_branch(&name, git2::BranchType::Remote)
-                    .ok()
-                    .and_then(|b| b.get().target())
-                    .ok_or_else(|| anyhow::format_err!("failed to find {refkind} `{s}`"))?
-            }
+        let result = match self {
+            // Resolve the commit pointed to by the tag.
+            Self::Tag(s) => repo.rev_parse(&format!("refs/remotes/origin/tags/{s}^0")),
```

Comment:
What does the `^0` signify here?

Reply:
`^0` is short for `^{commit}`, which recursively "peels" to the underlying commit object. e.g. if the tag points to another tag, which then points to a commit. I'm not 100% sure if we need to do this, but the previous code had calls to `peel(ObjectType::Commit)`, so this retains the original behavior. I'll add a more detailed comment explaining this.